### ISSUE TITLE
Requirements to upstream uvwasi integration

### ISSUE THREAD

**Issue by cjihrig** [cjihrig/wasm3#1]
Hi @vshymanskyy 👋 

I've done some [initial work to migrate wasm3's WASI implementation to use uvwasi](https://github.com/cjihrig/wasm3/commits/uvwasi). I've summarized the status of my work below. I'm interested in moving this forward, and can commit to helping maintain it in the future.

However, this is also turning out to be significantly more up front work than I anticipated. I'm curious what would be needed minimally to upstream this work, assuming that it would continue to be iterated on. Any help in the form of code or feedback would be appreciated.

Notes related to the code:

- [ ] I have ported the existing WASI calls to use uvwasi. The exceptions are `args_sizes_get()` and `args_get()` which can trivially be implemented without uvwasi. The reason I did not port them is because the runtime doesn't appear to be completely initialized at the time WASI is initialized. I've left a comment stating this in the code. I've also implemented some missing WASI functionality, such as working with environment variables.
- [ ] There are more WASI APIs that were not originally implemented in wasm3. I can work on adding them in the future, but this is already a substantial amount of work 😄  
- [ ] To properly use preopens, the functionality of the `--dir` CLI flag must be expanded. I've left a TODO comment about this as well. Currently, only the current working directory is preopened as the root directory in the sandbox. In the process I [found and addressed a bug in uvwasi](https://github.com/cjihrig/uvwasi/pull/105).
- [ ] The memory write back logic in `m3_wasi_unstable_fd_prestat_get()` still has a bug. The code is currently wrong, but allows WASI applications to at least run correctly. Changing the code to be correct causes issues. I've left a TODO comment regarding this.

Notes related to building:

- [ ] I'm currently integrating uvwasi and libuv via CMake's `FetchContent` feature. That would require a slight bump in the `cmake_minimum_required` version. It's not clear to me if this is the preferred way to integrate with the build. If it is not, what is?
- [ ] What impact, if any, do these dependencies have on projects that already include libuv, such as txiki.js.
- [ ] `-Werror=shadow` is currently disabled as it caused a problem compiling libuv. I'm guessing we'd want this re-enabled.

Notes related to testing:

- [ ] Currently, `test/wasi/test.wasm` is working. However, I had to recompile it. I did so using [this approach](https://github.com/nodejs/node/blob/83f7bb441fb647ccac88f4787ac0c308ebb146ab/test/wasi/README.md) (not the one described in the wasm3 docs). Any other relevant test files will also need to be regenerated, as the current files are based on an obsolete version of the WASI spec.

**Comment by vshymanskyy** [comment]
This is awesome, man! Here are some thoughts:
1. The existing `m3_api_wasi.c` implementation should be untouched, as it's still preferred in some environments, or if one needs just that. So we need to move the uvwasi implementation to `m3_api_uvwasi.c` (and guard it with `d_m3HasUVWASI` flag). This would be **the only** requirement for this to be integrated upstream.
2. `args_sizes_get`, `args_get` - no need to change.
3. The `--dir` flag was recently added, so it can be integrated. I think it would be OK to have a hardcoded limit of say 8 or 16 items.
4. `m3_wasi_unstable_fd_prestat_get` bug - this had to be investigated anyway. I suspect wrong alignment or mapping of the structure.
5. `txiki.js` - I think it'll just have to stick to the same version of `libuv` (shouldn't be a problem).
6. `FetchContent` - no objections. We can switch to using git submodules later on. `wasm3` will be restructured at some point.
7. `-Werror=shadow`. I'm not sure why this impacts `libuv` compilation, will have to dig into this. However I don't mind removing the flag temporarily.
8. Rebuilding the examples should not be a big deal, especially when https://github.com/wasienv/wasienv/issues/24 is resolved. But I think we could just link the same functions to the `wasi_unstable` namespace (at least `wasmer` is doing this, if I recall correctly).

Huge thanks for working on this :+1:

**Comment by cjihrig** [comment]
I've pushed up https://github.com/cjihrig/wasm3/commit/217df7e58d7a8e8f0c1c448ba11259d7dca08327, which should address a lot of these.

> `-Werror=shadow`. I'm not sure why this impacts `libuv` compilation, will have to dig into this.

This is an issue with libuv itself. I updated the cmake build to still enable it when libuv is not being built.

> But I think we could just link the same functions to the `wasi_unstable` namespace

I reverted my changes to the test file, and have uvwasi linking as `wasi_unstable`. However, even if wasm3 sticks with this approach, the function signatures are prone to change until WASI stabilizes. I don't think it's an issue now, but it will probably become one over time.

I think the only remaining thing for me to do is [release uvwasi 0.0.7](https://github.com/cjihrig/uvwasi/pull/106) and use that here. Right now it's using 0.0.6 but I patched it locally to include the [bug fix](https://github.com/cjihrig/uvwasi/pull/105) I mentioned previously.
